action/protocol/staking/vote_reviser.go

```diff
@@ -2,43 +2,61 @@ package staking
 
 import (
 	"math/big"
+	"slices"
 	"sort"
 
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
 
+	"github.com/iotexproject/iotex-core/action/protocol"
 	"github.com/iotexproject/iotex-core/blockchain/genesis"
 	"github.com/iotexproject/iotex-core/pkg/log"
 	"github.com/iotexproject/iotex-core/state"
 )
 
 // VoteReviser is used to recalculate candidate votes.
-type VoteReviser struct {
-	reviseHeights      []uint64
-	cache              map[uint64]CandidateList
-	c                  genesis.VoteWeightCalConsts
-	correctCandsHeight uint64
-}
+type (
+	VoteReviser struct {
+		cache map[uint64]CandidateList
+		cfg   ReviseConfig
+	}
+
+	ReviseConfig struct {
+		VoteWeight                  genesis.VoteWeightCalConsts
+		ReviseHeights               []uint64
+		CorrectCandsHeight          uint64
+		SelfStakeBucketReviseHeight uint64
+	}
+)
 
 // NewVoteReviser creates a VoteReviser.
-func NewVoteReviser(c genesis.VoteWeightCalConsts, correctCandsHeight uint64, reviseHeights ...uint64) *VoteReviser {
+func NewVoteReviser(cfg ReviseConfig) *VoteReviser {
+	// TODO: return error if cfg.CorrectSelfStakeBucketHeights is before hardfork height
 	return &VoteReviser{
-		reviseHeights:      reviseHeights,
-		cache:              make(map[uint64]CandidateList),
-		c:                  c,
-		correctCandsHeight: correctCandsHeight,
+		cfg:   cfg,
+		cache: make(map[uint64]CandidateList),
 	}
 }
 
 // Revise recalculate candidate votes on preset revising height.
-func (vr *VoteReviser) Revise(csm CandidateStateManager, height uint64) error {
+func (vr *VoteReviser) Revise(ctx protocol.FeatureCtx, csm CandidateStateManager, height uint64) error {
 	if !vr.isCacheExist(height) {
-		cands, err := vr.calculateVoteWeight(csm)
+		cands, _, err := newCandidateStateReader(csm.SM()).getAllCandidates()
+		switch {
+		case errors.Cause(err) == state.ErrStateNotExist:
+		case err != nil:
+			return err
+		}
+		cands, err = vr.reviseSelfStakeBuckets(ctx, csm, height, cands)
```

Comment:
revise self stake is an option for this time, not needed for every revise, so better to write like
```
if vr.shouldReviseSelfStakeBuckets(height) {
    cands, err = vr.reviseSelfStakeBuckets()
}
```

Reply:
also is it possible to only revise affected cands? like
```
if vr.shouldReviseSelfStakeBuckets(height) {
    affectedCands, err := vr.reviseSelfStakeBuckets()
    cands, err = vr.calculateVoteWeight(csm, height, affectedCands)
} else {
    // revise all cands
    cands, err = vr.calculateVoteWeight(csm, height, affectedCands)
}
```